doc/RFC-MMP.llnl.04.txt

```diff
@@ -0,0 +1,913 @@
+labels found contain MMP blocks,
+    "zpool [add|attach|replace|create] --force <vdev>"
+will fail with an appropriate message.
+
+If there are no labels on the device(s), the device will be partitioned as
+normal.
+
+=== Zpool labelclear ===
+
+Zpool labelclear zeroes the entire content of all the labels, including the
+uberblock rings.
+
+Zpool labelclear is left unaltered so that users have a way to clear MMP
+information from devices that are no longer part of an active pool, but which
+had MMP enabled (e.g. disks from a retired node are re-used in a new node).
+
+If zpool labelclear is used on a device which is in use by an open pool, the
+other node will detect the change to the on-disk MMP blocks, triggering an
+MMP fault.
+
+(XXX recovery options in this scenario depend on the MMP Fault options, which
+need to be settled.)
+
+=== Pool suspension ===
+
+When a pool recovers from suspension due to IO failures (in
+zio_vdev_resume_io()), we should reread all MMP blocks before resuming.  We
+should only resume if the MMP blocks were not changed based on their checksums.
+
+=== Feature Flag ===
+
+It would be useful to be able to turn MMP off when not using shared storage,
+since there is some performance cost.  It would also be useful for moving a
+pool to another node whose ZFS install does not support MMP.
+
+Implementing a feature flag would both allow the user to control MMP and
+provide a notification mechanism when importing a pool under another
+ZFS implementation.
+
+MMP is off by default.
+
+=== I/O Errors within MMP ===
+
+(XXX MMP code may encounter I/O errors.  Devices may fail.  MMP should be as
+tolerant of them as the rest of ZFS.  Are there design issues here?)
+
+=== HostID ===
+
+(XXX What do we do with host id?  Drop that code from ZFS entirely?  Or retain
+it as the fallback for cases where MMP is disabled?.  What are the implications
+for "zpool import --force"?)
+
+=== Misc ===
+^ Were any important operations missed, other than import/add/attach/online/replace?
+
+^ What about zpool split?
+
+^ What about zpool reguid?  zpool repoen?
+
+^ What about log devices?  l2arc devices?
+
+^ We could keep the checksum of every uberblock, not just the the ones in mmp
+slots.  Keeping them would be inexpensive; just read them all once at import
+time.  Then after that reading them before a sync would detect concurrent
+import even by a non-MMP node.  Without this, protection against a non-mmp node
+is weak.  Eventually one of the MMP blocks will be changed but that could take
+a long time.  Reading them in for that test would be the expensive part.  Since
+the pool should reflect that the MMP feature is enabled, it should not be
+possible to import the pool on a non-MMP aware node, so this should not be
+necessary.
+
+^ Jeff suggested to use bdev_strategy(?) for submitting the IO, to
+ reduce the IO latency as much as possible.  This will ensure no more
+ than 10 or so IOs are in front of the MMP block (about 1/10s or less).
```

Comment:
"10 or so" is defined by min(zfs_vdev_max_active, sum(zfs_vdev_*_max_active)), which is 35 by default.